src/org/traccar/api/BaseResource.java

```diff
@@ -29,4 +39,34 @@ protected long getUserId() {
         }
         return 0;
     }
+
+    protected void checkAndLinkPermission(LinkedHashMap<String, Long> entity, boolean link)
+            throws SQLException, ClassNotFoundException {
+        Iterator<String> iterator = entity.keySet().iterator();
+        String owner = iterator.next();
+        Class<?> ownerClass = DataManager.getClassByName(owner);
+        String property = iterator.next();
+        Class<?> propertyClass = DataManager.getClassByName(property);
+
+        long ownerId = entity.get(owner);
+        long propertyId = entity.get(property);
+
+        if (!link && ownerClass.equals(User.class)
+                && propertyClass.equals(Device.class)) {
+            if (getUserId() != ownerId) {
+                Context.getPermissionsManager().checkUser(getUserId(), ownerId);
+            } else {
+                Context.getPermissionsManager().checkAdmin(getUserId());
+            }
+        } else {
+            Context.getPermissionsManager().checkPermission(ownerClass, getUserId(), ownerId);
+        }
+        Context.getPermissionsManager().checkPermission(propertyClass, getUserId(), propertyId);
+
+        Context.getDataManager().linkObject(ownerClass, ownerId, propertyClass, propertyId, link);
+    }
+
+    protected void linkNewEntity(BaseModel entity) throws SQLException {
```

Comment:
But this method substitute 3 common values.  It is called many times, and only `entity` is changed